### PR TITLE
[DeepVision port] Adding CLIP + pretrained weights

### PR DESCRIPTION
**Pull request by DavidLandup0**
# What does this PR do?

As discussed in #1933 - setting up a draft PR for porting CLIP and associated layers into KCV. Draft PR for now with placeholder main model dump, layers and tests incoming soon. Will tag once ready for review.

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case. #1933
- [ ] Did you write any new necessary tests?
- [ ] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ianstenbit just tagging so you can follow the progress as it comes in. Otherwise, no need to spend time until it's un-drafted for review :)

### CODE REVIEW

File: keras_cv/models/clip/clip_tokenizer.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+
+@lru_cache()
+def default_bpe():
+    return os.path.join(
+        os.path.dirname(os.path.abspath(__file__)), "bpe_simple_vocab_16e6.txt.gz"
+    )
+
+
+class CLIPTokenizer(object):
```
Comment (jbischof) [inline keras_cv/models/clip/clip_tokenizer.py:18]: I think this if no where else is an excellent opportunity to import from KerasNLP. We already have a BPE tokenizer and downstream GPT2 tokenizer!

File: keras_cv/models/clip/clip_image_encoder.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import torch.nn.functional as F
+from tensorflow.keras import layers
+from torch import nn
+
+from deepvision.layers.clip_patching_and_embedding import CLIPPatchingAndEmbedding
+from deepvision.layers.residual_transformer_encoder import ResidualTransformerEncoder
+from deepvision.utils.utils import parse_model_inputs
+
+
+class __CLIPImageEncoderTF(tf.keras.Model):
```
Comment (jbischof) [inline keras_cv/models/clip/clip_image_encoder.py:18]: Remove the underscores in class names

File: keras_cv/models/clip/clip_image_encoder.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+from typing import Union
+
+import numpy as np
+import tensorflow as tf
+import torch
+import torch.nn.functional as F
+from tensorflow.keras import layers
+from torch import nn
+
+from deepvision.layers.clip_patching_and_embedding import CLIPPatchingAndEmbedding
```
Comment (jbischof) [inline keras_cv/models/clip/clip_image_encoder.py:13]: I'm a little confused. Why is this a TF only version? Why are we importing Torch modules?

**Review by jbischof — COMMENTED** [review]
Cool! Left some early comments

**Comment by DavidLandup0** [comment]
@jbischof you got in a bit early on the review 😅 
This is a dump of the files as they are. I'll remove the underscored, `torch` imports etc. and open it from draft onto a review in the next few days :)

Good idea on importing portions from KNLP!

File: keras_cv/models/clip/clip_tokenizer.py (hunk 1 above)
Comment (DavidLandup0) [reply keras_cv/models/clip/clip_tokenizer.py:18]: Agreed! I'll rely on those then :)
Thanks for the heads up!

File: keras_cv/models/clip/clip_image_encoder.py (hunk 2 above)
Comment (DavidLandup0) [reply keras_cv/models/clip/clip_image_encoder.py:18]: Yep, planned, they're used only internally in the other repo so they're underscored and another API is exposed. This'll be removed when the PR is ready for review

File: keras_cv/models/clip/clip_image_encoder.py (hunk 3 above)
Comment (DavidLandup0) [reply keras_cv/models/clip/clip_image_encoder.py:13]: Comment in the main thread 🙇

**Comment by jbischof** [comment]
Actually one complexity on borrowing the NLP tokenizer is that it is written in TF and cannot be ported to Keras Core due to tf_text dependency. So maybe you can borrow the Stable Diffusion one instead.

**Comment by divyashreepathihalli** [comment]
@DavidLandup0 Thank you for this PR! are you still working on this PR? I would create a feature branch for adding the CLIP model and you can add your changes to that.

**Comment by divyashreepathihalli** [comment]
@DavidLandup0 I have cherry picked your changes to this branch - https://github.com/keras-team/keras-cv/tree/CLIP

**Pull request closed without merging**